File: architecture/testing/across-languages/python/service.py
"""Тонкий слой над доменом — то, что тестируем дублёрами."""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Protocol

import pricing


@dataclass(frozen=True)
class Order:
    """Сохранённый заказ: total_cent — до скидки, price_cent — к оплате."""

    id: str
    user_id: str
    total_cent: int
    price_cent: int


class Store(Protocol):
    """Граница нашего кода: хранилище заказов."""

    def save(self, o: Order) -> None: ...

    def by_user(self, user_id: str) -> list[Order]: ...


class Notifier(Protocol):
    """Вторая граница: отправка уведомления."""

    def notify(self, user_id: str, text: str) -> None: ...


class OrderService:
    def __init__(self, store: Store, notifier: Notifier) -> None:
        self._store = store
        self._notifier = notifier

    def create(self, id: str, user_id: str, total_cents: int) -> Order:
        """Считает цену, сохраняет заказ, уведомляет пользователя.

        Переменная окружения DISCOUNT_BUG=1 включает намеренный дефект
        (скидка теряется) — им проверяем, кто из дублёров его ловит.
        """
        if total_cents < 0:
            raise ValueError(f"сумма заказа отрицательна: {total_cents}")

        if os.environ.get("DISCOUNT_BUG") == "1":
            p = total_cents  # БАГ: скидка потеряна
        else:
            p = pricing.price(total_cents)

        o = Order(id=id, user_id=user_id, total_cent=total_cents, price_cent=p)
        self._store.save(o)
        try:
            # Уведомление не критично: заказ уже сохранён.
            self._notifier.notify(user_id, f"Заказ {id} принят, к оплате {p} коп.")
        except Exception:  # noqa: BLE001 — намеренно проглатываем
            pass
        return o

File: architecture/testing/across-languages/go/service/service.go
// Package service — тонкий слой над доменом: посчитать цену, сохранить заказ,
// уведомить. Нужен, чтобы показать ПРИЁМ 3 — дублёров (мок/стаб/фейк) — на
// одинаковой задаче в трёх языках.
package service

import (
	"context"
	"fmt"

	"tech.khorost/across-languages-go/pricing"
)

// Order — сохранённый заказ.
type Order struct {
	ID        string
	UserID    string
	TotalCent int64 // сумма до скидки
	PriceCent int64 // к оплате
}

// Store — граница нашего кода: хранилище заказов. Порт, а не деталь: реализаций
// две — Postgres в проде и фейк в тестах.
type Store interface {
	Save(ctx context.Context, o Order) error
	ByUser(ctx context.Context, userID string) ([]Order, error)
}

// Notifier — вторая граница: отправка уведомления.
type Notifier interface {
	Notify(ctx context.Context, userID, text string) error
}

// OrderService — то, что тестируем.
type OrderService struct {
	Store    Store
	Notifier Notifier
}

// Create считает цену, сохраняет заказ и уведомляет пользователя.
func (s OrderService) Create(ctx context.Context, id, userID string, totalCents int64) (Order, error) {
	if totalCents < 0 {
		return Order{}, fmt.Errorf("сумма заказа отрицательна: %d", totalCents)
	}
	price := pricing.Price(totalCents)
	if discountBug {
		price = totalCents // БАГ (только под тегом bug): скидка потеряна
	}
	o := Order{
		ID:        id,
		UserID:    userID,
		TotalCent: totalCents,
		PriceCent: price,
	}
	if err := s.Store.Save(ctx, o); err != nil {
		return Order{}, fmt.Errorf("сохранение заказа: %w", err)
	}
	// Уведомление — не критично: заказ уже сохранён, ошибку не эскалируем.
	_ = s.Notifier.Notify(ctx, userID, fmt.Sprintf("Заказ %s принят, к оплате %d коп.", id, o.PriceCent))
	return o, nil
}
